File: kotlin/functions/class_example.kt
open class Person(var name: String, var age: Int) {
    fun greet(): String {
        return "Hello, I'm $name and I'm $age years old"
    }

    fun birthday() {
        age++
        println("Happy birthday! Now $age years old")
    }

    override fun toString(): String {
        return "Person(name=$name, age=$age)"
    }
}

class Employee(name: String, age: Int, val jobTitle: String) : Person(name, age) {
    fun work(): String {
        return "$name is working as a $jobTitle"
    }

    override fun toString(): String {
        return "Employee(name=$name, age=$age, job=$jobTitle)"
    }
}

// Data class example
data class Product(val name: String, val price: Double, val quantity: Int)

fun main() {
    val person = Person("Alice", 30)
    println(person.greet())
    person.birthday()
    println(person)

    println()

    val employee = Employee("Bob", 25, "Software Engineer")
    println(employee.greet())
    println(employee.work())
    println(employee)

    println()

    val product = Product("Laptop", 999.99, 5)
    println("Product: $product")

    // Data class features
    val product2 = product.copy(quantity = 10)
    println("Copied product: $product2")
}

File: cpp/functions/class_example.cpp
#include <iostream>
#include <string>

class Person {
protected:
    std::string name;
    int age;

public:
    Person(const std::string& name, int age) : name(name), age(age) {}

    std::string greet() const {
        return "Hello, I'm " + name + " and I'm " + std::to_string(age) + " years old";
    }

    void birthday() {
        age++;
        std::cout << "Happy birthday! Now " << age << " years old" << std::endl;
    }

    virtual void display() const {
        std::cout << "Person(name=" << name << ", age=" << age << ")" << std::endl;
    }

    virtual ~Person() = default;
};

class Employee : public Person {
private:
    std::string jobTitle;

public:
    Employee(const std::string& name, int age, const std::string& jobTitle)
        : Person(name, age), jobTitle(jobTitle) {}

    std::string work() const {
        return name + " is working as a " + jobTitle;
    }

    void display() const override {
        std::cout << "Employee(name=" << name << ", age=" << age
                  << ", job=" << jobTitle << ")" << std::endl;
    }
};

int main() {
    Person person("Alice", 30);
    std::cout << person.greet() << std::endl;
    person.birthday();
    person.display();

    std::cout << std::endl;

    Employee employee("Bob", 25, "Software Engineer");
    std::cout << employee.greet() << std::endl;
    std::cout << employee.work() << std::endl;
    employee.display();

    return 0;
}
